pandas/io/parsers.py

```diff
@@ -1894,23 +1899,52 @@ def __init__(self, src, **kwds):
 
         ParserBase.__init__(self, kwds)
 
-        encoding = kwds.get("encoding")
-
-        # parsers.TextReader doesn't support compression dicts
-        if isinstance(kwds.get("compression"), dict):
-            kwds["compression"] = kwds["compression"]["method"]
-
-        if kwds.get("compression") is None and encoding:
-            if isinstance(src, str):
-                src = open(src, "rb")
-                self.handles.append(src)
-
-            # Handle the file object with universal line mode enabled.
-            # We will handle the newline character ourselves later on.
-            if hasattr(src, "read") and not hasattr(src, "encoding"):
-                src = TextIOWrapper(src, encoding=encoding, newline="")
-
-            kwds["encoding"] = "utf-8"
+        src = stringify_path(src)
+        if (
+            kwds.get("memory_map", False)
+            and isinstance(src, str)
+            and get_compression_method(kwds.get("compression", None))[0] is None
+        ):
+            # memory-mapped files are directly handled by the TextReader.
+            # compression and file-object were never supported by it.
+            self.source = src
+            self.handles = []
+        else:
+            if kwds.get("memory_map", False):
```

Comment:
isn't this also true in the above if?

Reply:
yes, if the conditions for `memory_map=True` aren't met (string + no compression), it falls back to `memory_map=False`. Or it could raise an error, I'm not sure what is preferable.

Reply:
I think we should refactor the `if-else` to have only one `if kwds.get("memory_map", False)`

That would improve readability IMO.

Reply:
yes, that is now more readable :)